Mypy failure: 'error: <nothing> has no attribute "pipe"'

This is happening on the main build: e.g. https://github.com/pystatgen/sgkit/runs/5312420654?check_suite_focus=true

```
Check for merge conflicts................................................Passed
Debug Statements (Python)................................................Passed
Mixed line ending........................................................Passed
Check for case conflicts.................................................Passed
Check Yaml...............................................................Passed
isort....................................................................Passed
black....................................................................Passed
flake8...................................................................Passed
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

sgkit/tests/test_pc_relate.py:105: error: <nothing> has no attribute "pipe"  [attr-defined]
sgkit/tests/test_pc_relate.py:143: error: <nothing> has no attribute "pipe"  [attr-defined]
```

I tried updating mypy to the latest version (v0.931, we are currently on v0.782), but this produces lots of errors:

```
setup.py:2: error: Skipping analyzing "setuptools": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/io/vcf/__init__.py:1: error: Cannot find implementation or library stub for module named "pytest"  [import]
docs/extensions/typed_returns.py:10: error: Cannot find implementation or library stub for module named "sphinx.application"  [import]
docs/extensions/typed_returns.py:11: error: Cannot find implementation or library stub for module named "sphinx.ext.napoleon"  [import]
validation/gwas/method/regenie/tasks.py:8: error: Library stubs not installed for "yaml" (or incompatible with Python 3.8)  [import]
validation/gwas/method/regenie/tasks.py:9: error: Cannot find implementation or library stub for module named "invoke"  [import]
validation/gwas/method/regenie/tasks.py:22: note: (Skipping most remaining errors due to unresolved imports or missing stubs; fix these first)
validation/gwas/method/regenie/sgkit_zarr.py:8: error: Cannot find implementation or library stub for module named "fire"  [import]
validation/gwas/method/regenie/sgkit_zarr.py:9: error: Library stubs not installed for "yaml" (or incompatible with Python 3.8)  [import]
validation/gwas/method/regenie/sgkit_zarr.py:10: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
validation/gwas/method/regenie/sgkit_zarr.py:11: error: Cannot find implementation or library stub for module named "sgkit_plink"  [import]
validation/gwas/method/regenie/glow_wgr.py:14: error: Cannot find implementation or library stub for module named "fire"  [import]
validation/gwas/method/regenie/glow_wgr.py:15: error: Cannot find implementation or library stub for module named "glow"  [import]
validation/gwas/method/regenie/glow_wgr.py:17: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
validation/gwas/method/regenie/glow_wgr.py:18: error: Library stubs not installed for "pkg_resources" (or incompatible with Python 3.8)  [import]
validation/gwas/method/regenie/glow_wgr.py:18: note: Hint: "python3 -m pip install types-setuptools"
validation/gwas/method/regenie/glow_wgr.py:19: error: Cannot find implementation or library stub for module named "pyspark.sql.functions"  [import]
validation/gwas/method/regenie/glow_wgr.py:19: error: Cannot find implementation or library stub for module named "pyspark"  [import]
validation/gwas/method/regenie/glow_wgr.py:19: error: Cannot find implementation or library stub for module named "pyspark.sql"  [import]
validation/gwas/method/regenie/glow_wgr.py:20: error: Library stubs not installed for "yaml" (or incompatible with Python 3.8)  [import]
validation/gwas/method/regenie/glow_wgr.py:22: error: Cannot find implementation or library stub for module named "glow.wgr.functions"  [import]
validation/gwas/method/regenie/glow_wgr.py:27: error: Cannot find implementation or library stub for module named "glow.wgr.linear_model"  [import]
validation/gwas/method/regenie/glow_wgr.py:29: error: Cannot find implementation or library stub for module named "pyspark.sql.session"  [import]
sgkit/typing.py:4: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/typing.py:4: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
validation/gwas/method/hwe/tasks.py:10: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
validation/gwas/method/hwe/tasks.py:11: error: Cannot find implementation or library stub for module named "invoke"  [import]
validation/gwas/method/regenie/hail_sim.py:9: error: Cannot find implementation or library stub for module named "fire"  [import]
validation/gwas/method/regenie/hail_sim.py:10: error: Cannot find implementation or library stub for module named "hail"  [import]
validation/gwas/method/regenie/hail_sim.py:12: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
validation/gwas/method/regenie/hail_sim.py:13: error: Library stubs not installed for "yaml" (or incompatible with Python 3.8)  [import]
validation/gwas/method/regenie/hail_sim.py:13: note: Hint: "python3 -m pip install types-PyYAML"
validation/gwas/method/regenie/hail_sim.py:13: note: (or run "mypy --install-types" to install all missing stub packages)
sgkit/io/vcf/utils.py:9: error: Skipping analyzing "fsspec": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcf/utils.py:10: error: Cannot find implementation or library stub for module named "yarl"  [import]
sgkit/distance/metrics.py:12: error: Skipping analyzing "numba": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/cohorts.py:4: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_cohorts.py:2: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_cohorts.py:3: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/test_utils.py:5: error: Skipping analyzing "fsspec": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/io/vcf/test_utils.py:6: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/test_utils.py:7: error: Cannot find implementation or library stub for module named "callee.strings"  [import]
sgkit/display.py:3: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/utils.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/utils.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/distance/api.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/distance/api.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/dataset.py:4: error: Skipping analyzing "fsspec": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_stats_utils.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_stats_utils.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_stats_utils.py:5: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_stats_utils.py:7: error: Cannot find implementation or library stub for module named "hypothesis"  [import]
sgkit/utils.py:5: error: Skipping analyzing "numba": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/preprocessing.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/preprocessing.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/preprocessing.py:6: error: Skipping analyzing "sklearn.base": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/aggregation.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/aggregation.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/aggregation.py:6: error: Skipping analyzing "numba": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/regenie.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/regenie.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/conversion.py:1: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/conversion.py:1: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/conversion.py:3: error: Skipping analyzing "numba": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/utils.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/utils.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/utils.py:4: error: Skipping analyzing "dask.dataframe": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/utils.py:6: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/association.py:4: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/association.py:4: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/window.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/window.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/pc_relate.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/pc_relate.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/pca.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/pca.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/pca.py:6: error: Skipping analyzing "dask_ml.decomposition": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/pca.py:7: error: Skipping analyzing "sklearn.base": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/pca.py:8: error: Skipping analyzing "sklearn.pipeline": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/popgen.py:5: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/popgen.py:5: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/popgen.py:7: error: Skipping analyzing "numba": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/popgen.py:746: error: Unused "type: ignore" comment
sgkit/stats/hwe.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/hwe.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/hwe.py:5: error: Skipping analyzing "numba": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/ibs.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/ibs.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/ld.py:4: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/ld.py:5: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/ld.py:6: error: Skipping analyzing "dask.dataframe": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/ld.py:8: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/stats/ld.py:10: error: Skipping analyzing "numba": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcfzarr_reader.py:16: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcfzarr_reader.py:17: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcfzarr_reader.py:18: error: Skipping analyzing "numcodecs": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcfzarr_reader.py:21: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcfzarr_reader.py:22: error: Skipping analyzing "dask.delayed": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcfzarr_reader.py:23: error: Skipping analyzing "dask.optimization": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcfzarr_reader.py:24: error: Skipping analyzing "fsspec": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcfzarr_reader.py:25: error: Skipping analyzing "scipy.special": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/__init__.py:1: error: Library stubs not installed for "pkg_resources" (or incompatible with Python 3.8)  [import]
sgkit/io/vcf/vcf_partition.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcf/vcf_partition.py:4: error: Skipping analyzing "fsspec": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcf/vcf_partition.py:6: error: Cannot find implementation or library stub for module named "cyvcf2"  [import]
sgkit/tests/test_distance.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_distance.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_distance.py:5: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_distance.py:6: error: Skipping analyzing "numba": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_distance.py:7: error: Skipping analyzing "scipy.spatial.distance": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_hwe.py:5: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_hwe.py:6: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_regenie.py:6: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_regenie.py:6: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_regenie.py:8: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_regenie.py:9: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_regenie.py:11: error: Library stubs not installed for "yaml" (or incompatible with Python 3.8)  [import]
sgkit/tests/test_regenie.py:12: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_regenie.py:14: error: Cannot find implementation or library stub for module named "hypothesis"  [import]
sgkit/tests/test_regenie.py:16: error: Cannot find implementation or library stub for module named "hypothesis.extra.numpy"  [import]
sgkit/tests/test_regenie.py:17: error: Cannot find implementation or library stub for module named "hypothesis.strategies"  [import]
sgkit/tests/test_pca.py:3: error: Cannot find implementation or library stub for module named "allel"  [import]
sgkit/tests/test_pca.py:4: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_pca.py:4: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_pca.py:6: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/test_dataset.py:3: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_window.py:3: error: Cannot find implementation or library stub for module named "allel"  [import]
sgkit/tests/test_window.py:4: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_window.py:4: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_window.py:6: error: Cannot find implementation or library stub for module named "pytest"  [import]
docs/conf.py:19: error: Skipping analyzing "dask.dataframe": module is installed, but missing library stubs or py.typed marker  [import]
docs/conf.py:19: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
docs/conf.py:21: error: Cannot find implementation or library stub for module named "sphinx.util"  [import]
sgkit/tests/test_display.py:4: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_aggregation.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_aggregation.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_aggregation.py:5: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_vcfzarr_reader.py:1: error: Cannot find implementation or library stub for module named "allel"  [import]
sgkit/tests/test_vcfzarr_reader.py:3: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_vcfzarr_reader.py:5: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_variables.py:2: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_conversion.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_conversion.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_conversion.py:5: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_utils.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_utils.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_utils.py:5: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_utils.py:7: error: Cannot find implementation or library stub for module named "hypothesis"  [import]
sgkit/tests/test_utils.py:9: error: Cannot find implementation or library stub for module named "hypothesis.extra.numpy"  [import]
sgkit/tests/test_testing.py:3: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/io/plink/plink_reader.py:5: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/plink/plink_reader.py:5: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/plink/plink_reader.py:6: error: Skipping analyzing "dask.dataframe": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/plink/plink_reader.py:8: error: Cannot find implementation or library stub for module named "bed_reader"  [import]
sgkit/tests/test_ld.py:3: error: Cannot find implementation or library stub for module named "allel"  [import]
sgkit/tests/test_ld.py:4: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_ld.py:4: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_ld.py:7: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_ld.py:8: error: Skipping analyzing "dask.dataframe": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_ld.py:9: error: Cannot find implementation or library stub for module named "hypothesis"  [import]
sgkit/tests/test_ld.py:11: error: Cannot find implementation or library stub for module named "hypothesis.extra.numpy"  [import]
sgkit/tests/test_mis.py:4: error: Skipping analyzing "networkx": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_mis.py:6: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_mis.py:7: error: Skipping analyzing "toolz": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/io/test_utils.py:2: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/test_utils.py:3: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/bgen/bgen_reader.py:18: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/bgen/bgen_reader.py:19: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/bgen/bgen_reader.py:20: error: Skipping analyzing "dask.dataframe": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/bgen/bgen_reader.py:22: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/bgen/bgen_reader.py:24: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/bgen/bgen_reader.py:25: error: Cannot find implementation or library stub for module named "cbgen"  [import]
sgkit/io/bgen/bgen_reader.py:26: error: Cannot find implementation or library stub for module named "rechunker"  [import]
sgkit/tests/test_preprocessing.py:3: error: Cannot find implementation or library stub for module named "allel.stats.preprocessing"  [import]
sgkit/tests/test_preprocessing.py:3: error: Cannot find implementation or library stub for module named "allel"  [import]
sgkit/tests/test_preprocessing.py:3: error: Cannot find implementation or library stub for module named "allel.stats"  [import]
sgkit/tests/test_preprocessing.py:4: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_preprocessing.py:4: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_preprocessing.py:6: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_pc_relate.py:2: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_pc_relate.py:3: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_pc_relate.py:5: error: Cannot find implementation or library stub for module named "hypothesis"  [import]
sgkit/tests/test_pc_relate.py:6: error: Cannot find implementation or library stub for module named "hypothesis.extra.numpy"  [import]
sgkit/io/vcf/vcf_reader.py:19: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcf/vcf_reader.py:20: error: Skipping analyzing "fsspec": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/io/vcf/vcf_reader.py:23: error: Cannot find implementation or library stub for module named "cyvcf2"  [import]
sgkit/io/vcf/vcf_reader.py:24: error: Skipping analyzing "numcodecs": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_ibs.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_ibs.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
sgkit/tests/test_ibs.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_ibs.py:5: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/test_vcf_reader.py:4: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/test_vcf_reader.py:6: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/io/vcf/test_vcf_reader.py:7: error: Skipping analyzing "numcodecs": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/io/vcf/test_vcf_roundtrip.py:25: error: Cannot find implementation or library stub for module named "allel"  [import]
sgkit/tests/io/vcf/test_vcf_roundtrip.py:27: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/test_csi.py:1: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/test_csi.py:2: error: Cannot find implementation or library stub for module named "cyvcf2"  [import]
sgkit/tests/io/vcf/vcf_writer.py:8: error: Cannot find implementation or library stub for module named "cyvcf2"  [import]
sgkit/tests/test_popgen.py:3: error: Cannot find implementation or library stub for module named "allel"  [import]
sgkit/tests/test_popgen.py:4: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_popgen.py:4: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_popgen.py:7: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/test_vcf_partition.py:1: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_association.py:5: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_association.py:5: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_association.py:7: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_association.py:8: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/test_association.py:10: error: Skipping analyzing "zarr": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/test_association.py:37: error: Cannot find implementation or library stub for module named "statsmodels.api"  [import]
sgkit/tests/test_association.py:37: error: Cannot find implementation or library stub for module named "statsmodels"  [import]
sgkit/tests/test_association.py:38: error: Cannot find implementation or library stub for module named "statsmodels.regression.linear_model"  [import]
sgkit/tests/io/bgen/test_bgen_reader.py:6: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/test_tbi.py:1: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/plink/test_plink_reader.py:2: error: Cannot find implementation or library stub for module named "pytest"  [import]
sgkit/tests/io/vcf/vcf_generator.py:9: error: Skipping analyzing "scipy.special": module is installed, but missing library stubs or py.typed marker  [import]
sgkit/tests/io/vcf/test_vcf_lossless_conversion.py:1: error: Cannot find implementation or library stub for module named "pytest"  [import]
validation/gwas/method/pc_relate/validate_pc_relate.py:3: error: Skipping analyzing "dask.array": module is installed, but missing library stubs or py.typed marker  [import]
validation/gwas/method/pc_relate/validate_pc_relate.py:3: error: Skipping analyzing "dask": module is installed, but missing library stubs or py.typed marker  [import]
validation/gwas/method/pc_relate/validate_pc_relate.py:5: error: Skipping analyzing "pandas": module is installed, but missing library stubs or py.typed marker  [import]
Found 215 errors in 72 files (checked 105 source files)
```

We should probably move to the later version and fix these issues.

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -22,8 +22,8 @@ repos:
       - id: flake8
         language_version: python3
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.782
+    rev: v0.931
     hooks:
       - id: mypy
-        args: ["--strict", "--show-error-codes"]
-        additional_dependencies: ["numpy>=1.21.2", "xarray", "dask[array]", "scipy", "typing-extensions", "zarr", "numba", "dask-ml", "networkx"]
+        args: ["--show-error-codes", "--ignore-missing-imports"]
+        additional_dependencies: ["numpy>=1.21.2", "xarray", "dask[array]", "scipy", "typing-extensions", "zarr", "numba", "dask-ml", "networkx", "types-PyYAML"]

--- a/requirements.txt
+++ b/requirements.txt
@@ -6,7 +6,7 @@ dask-ml
 scipy
 typing-extensions
 numba
-zarr
+zarr >= 2.10.0, < 2.11.0
 fsspec != 2021.6.*
 scikit-learn
 pandas < 1.4.0

--- a/setup.cfg
+++ b/setup.cfg
@@ -32,7 +32,7 @@ install_requires =
     distributed >= 2021.10.0, <= 2022.01.0
     dask-ml
     scipy
-    zarr
+    zarr >= 2.10.0, < 2.11.0
     numba
     typing-extensions
     fsspec != 2021.6.*

--- a/sgkit/__init__.py
+++ b/sgkit/__init__.py
@@ -1,4 +1,4 @@
-from pkg_resources import DistributionNotFound, get_distribution
+from pkg_resources import DistributionNotFound, get_distribution  # type: ignore[import]
 
 from .display import display_genotypes
 from .distance.api import pairwise_distance

--- a/sgkit/io/bgen/bgen_reader.py
+++ b/sgkit/io/bgen/bgen_reader.py
@@ -145,7 +145,7 @@ class BgenReader:
                     res = np.zeros((len(all_vaddr), len(probs), 3), dtype=self.dtype)
                 res[i] = probs
             res = res[..., idx[2]]  # type: ignore[index]
-            return np.squeeze(res, axis=squeeze_dims)
+            return np.squeeze(res, axis=squeeze_dims)  # type: ignore[arg-type]
 
 
 def _split_alleles(allele_ids: bytes) -> List[bytes]:
@@ -520,7 +520,7 @@ def rechunk_bgen(
 
     zarr.consolidate_metadata(output)
 
-    ds: Dataset = xr.open_zarr(output, concat_characters=False)  # type: ignore[no-untyped-call]
+    ds = xr.open_zarr(output, concat_characters=False)  # type: ignore[no-untyped-call]
     if pack:
         ds = unpack_variables(ds)
 

--- a/sgkit/io/utils.py
+++ b/sgkit/io/utils.py
@@ -132,9 +132,9 @@ def _slice_zarrs(
 ) -> ArrayLike:
     """Slice concatenated zarrs by locs"""
     # convert array locations to slices
-    locs = [slice(*loc) for loc in locs]
+    locs = [slice(*loc) for loc in locs]  # type: ignore[misc]
     # determine which zarr files are needed
-    start, stop = locs[0].start, locs[0].stop  # stack on first axis
+    start, stop = locs[0].start, locs[0].stop  # type: ignore[attr-defined] # stack on first axis
     i0 = _zarr_index(offsets, start)
     i1 = _zarr_index(offsets, stop)
     if i0 == i1:  # within a single zarr file

--- a/sgkit/stats/pca.py
+++ b/sgkit/stats/pca.py
@@ -126,12 +126,14 @@ def pca_stats(ds: Dataset, est: BaseEstimator, *, merge: bool = True) -> Dataset
             _get(est, "explained_variance_ratio_"),
         ),
     }
-    new_ds = Dataset({k: v for k, v in new_ds.items() if v[1] is not None})
+    new_ds = Dataset({k: v for k, v in new_ds.items() if v[1] is not None})  # type: ignore[assignment]
     if "sample_pca_component" in new_ds and "sample_pca_explained_variance" in new_ds:
         new_ds[variables.sample_pca_loading] = new_ds[
             variables.sample_pca_component
-        ] * np.sqrt(new_ds[variables.sample_pca_explained_variance].data)
-    return conditional_merge_datasets(ds, variables.validate(new_ds), merge)
+        ] * np.sqrt(
+            new_ds[variables.sample_pca_explained_variance].data  # type: ignore[attr-defined]
+        )
+    return conditional_merge_datasets(ds, variables.validate(new_ds), merge)  # type: ignore[call-overload]
 
 
 def pca(

--- a/sgkit/stats/popgen.py
+++ b/sgkit/stats/popgen.py
@@ -265,7 +265,7 @@ def divergence(
     n_variants = ds.dims["variants"]
     n_cohorts = ds.dims["cohorts"]
     ac = da.asarray(ac)
-    shape = (ac.chunks[0], n_cohorts, n_cohorts)
+    shape = (ac.chunks[0], n_cohorts, n_cohorts)  # type: ignore[index]
     d = da.map_blocks(_divergence, ac, chunks=shape, dtype=np.float64)
     assert_array_shape(d, n_variants, n_cohorts, n_cohorts)
 
@@ -741,7 +741,7 @@ def pbs(
 
     # calculate PBS triples
     t = da.asarray(t)
-    shape = (t.chunks[0], n_cohorts, n_cohorts, n_cohorts)
+    shape = (t.chunks[0], n_cohorts, n_cohorts, n_cohorts)  # type: ignore[attr-defined]
 
     cohorts = cohorts or list(itertools.combinations(range(n_cohorts), 3))  # type: ignore
     ct = _cohorts_to_array(cohorts, ds.indexes.get("cohorts_0", None))
@@ -763,23 +763,23 @@ N_GARUD_H_STATS = 4  # H1, H12, H123, H2/H1
 def _Garud_h(haplotypes: ArrayLike) -> ArrayLike:
     # find haplotype counts (sorted in descending order)
     counts = sorted(collections.Counter(haplotypes.tolist()).values(), reverse=True)
-    counts = np.array(counts)
+    counts = np.array(counts)  # type: ignore[assignment]
 
     # find haplotype frequencies
     n = haplotypes.shape[0]
-    f = counts / n
+    f = counts / n  # type: ignore[operator]
 
     # compute H1
     h1 = np.sum(f ** 2)
 
     # compute H12
-    h12 = np.sum(f[:2]) ** 2 + np.sum(f[2:] ** 2)
+    h12 = np.sum(f[:2]) ** 2 + np.sum(f[2:] ** 2)  # type: ignore[index]
 
     # compute H123
-    h123 = np.sum(f[:3]) ** 2 + np.sum(f[3:] ** 2)
+    h123 = np.sum(f[:3]) ** 2 + np.sum(f[3:] ** 2)  # type: ignore[index]
 
     # compute H2/H1
-    h2 = h1 - f[0] ** 2
+    h2 = h1 - f[0] ** 2  # type: ignore[index]
     h2_h1 = h2 / h1
 
     return np.array([h1, h12, h123, h2_h1])

--- a/sgkit/stats/regenie.py
+++ b/sgkit/stats/regenie.py
@@ -151,7 +151,7 @@ def ridge_regression(
             # https://github.com/projectglow/glow/issues/266
             diag[:n_zero_reg] = 1
         diags.append(np.diag(diag))  # type: ignore[no-untyped-call]
-    diags = np.stack(diags)
+    diags = np.stack(diags)  # type: ignore[assignment]
     B = np.linalg.inv(XtX + diags) @ XtY  # type: ignore[no-untyped-call]
     B = B.astype(dtype or XtX.dtype)
     assert_array_shape(B, n_alpha, n_obs, n_outcome)
@@ -363,8 +363,8 @@ def _stage_2(
             alphas = get_alphas(n_variant_block * n_alpha_1)
     n_alpha_2 = alphas.size
 
-    YR = []
-    BR = []
+    YR_list = []
+    BR_list = []
     for i in range(n_outcome):
         # Slice and reshape to new 2D covariate matrix;
         # The order of raveling in trailing dimensions is important
@@ -382,11 +382,11 @@ def _stage_2(
         BB, YPB = _ridge_regression_cv(XPB, YB, alphas, n_zero_reg=n_covar)[-2:]
         assert_array_shape(BB, n_alpha_2, n_sample_block * n_indvar, 1)
         assert_array_shape(YPB, n_alpha_2, n_sample, 1)
-        BR.append(BB)
-        YR.append(YPB)
+        BR_list.append(BB)
+        YR_list.append(YPB)
 
     # Concatenate predictions along outcome dimension
-    YR = da.concatenate(YR, axis=2)
+    YR = da.concatenate(YR_list, axis=2)
     assert_block_shape(YR, 1, n_sample_block, n_outcome)
     assert_chunk_shape(YR, n_alpha_2, sample_chunks[0], 1)
     assert_array_shape(YR, n_alpha_2, n_sample, n_outcome)
@@ -399,7 +399,7 @@ def _stage_2(
     assert YR.shape[1:] == Y.T.shape
 
     # Concatenate betas along outcome dimension
-    BR = da.concatenate(BR, axis=2)
+    BR = da.concatenate(BR_list, axis=2)
     assert_block_shape(BR, 1, n_sample_block, n_outcome)
     assert_chunk_shape(BR, n_alpha_2, n_indvar, 1)
     assert_array_shape(BR, n_alpha_2, n_sample_block * n_indvar, n_outcome)

--- a/sgkit/tests/test_association.py
+++ b/sgkit/tests/test_association.py
@@ -146,14 +146,14 @@ def ds() -> Dataset:
 def _sm_statistics(
     ds: Dataset, i: int, add_intercept: bool
 ) -> RegressionResultsWrapper:
-    X = []
+    X_list = []
     # Make sure first independent variable is variant
-    X.append(ds["dosage"].values[i])
+    X_list.append(ds["dosage"].values[i])
     for v in [c for c in list(ds.keys()) if c.startswith("covar_")]:
-        X.append(ds[v].values)
+        X_list.append(ds[v].values)
     if add_intercept:
-        X.append(np.ones(ds.dims["samples"]))
-    X = np.stack(X).T
+        X_list.append(np.ones(ds.dims["samples"]))
+    X = np.stack(X_list).T
     y = ds[f"trait_{i}"].values
 
     return sm.OLS(y, X, hasconst=True).fit()

--- a/sgkit/tests/test_distance.py
+++ b/sgkit/tests/test_distance.py
@@ -80,7 +80,7 @@ def create_distance_matrix(
             chunks,
             device,
             metric,
-            marks=pytest.mark.gpu if device == "gpu" else "",
+            marks=pytest.mark.gpu if device == "gpu" else "",  # type: ignore[arg-type]
         )
         for shape in [(30, 30), (15, 30), (30, 15)]
         for chunks in [(10, 10), (5, 10), (10, 5)]

--- a/validation/gwas/method/regenie/glow_wgr.py
+++ b/validation/gwas/method/regenie/glow_wgr.py
@@ -1,3 +1,4 @@
+# type: ignore
 import os
 import sys
 

